src/core/lib/gprpp/dual_ref_counted.h

```diff
@@ -0,0 +1,342 @@
+//
+// Copyright 2020 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef GRPC_CORE_LIB_GPRPP_DUAL_REF_COUNTED_H
+#define GRPC_CORE_LIB_GPRPP_DUAL_REF_COUNTED_H
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/atm.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <atomic>
+#include <cassert>
+#include <cinttypes>
+
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/gprpp/atomic.h"
+#include "src/core/lib/gprpp/debug_location.h"
+#include "src/core/lib/gprpp/orphanable.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+
+namespace grpc_core {
+
+// DualRefCounted is an interface for reference-counted objects with two
+// classes of refs: strong refs (usually just called "refs") and weak refs.
+// Each class of refs can be incremented and decremented independently.
+//
+// Objects start with 1 strong ref and 0 weak refs at instantiation.
+// When the strong refcount reaches 0, the object's Orphan() method is called.
+// When the weak refcount reaches 0, the object is destroyed.
+//
+// This will be used by CRTP (curiously-recurring template pattern), e.g.:
+//   class MyClass : public RefCounted<MyClass> { ... };
+template <typename Child>
+class DualRefCounted : public Orphanable {
```

Comment:
This seems like it doesn't bring a huge amount of benefit, since it's hard to image how this would be misused.  But I also don't see any harm.

However, when I tried this, it generated a ton of compiler errors, which seem to stem from this:

```
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/type_traits:1302:66: error: incomplete type 'grpc_core::testing::(anonymous namespace)::Foo' used in type trait expression
    : public integral_constant<bool, __is_base_of(_Base, _Derived)>
                                                                 ^
./src/core/lib/gprpp/dual_ref_counted.h:55:22: note: in instantiation of template class 'std::is_base_of<grpc_core::DualRefCounted<grpc_core::testing::(anonymous namespace)::Foo>, grpc_core::testing::(anonymous namespace)::Foo>' requested here
  static_assert(std::is_base_of<DualRefCounted<Child>, Child>::value,
                     ^
test/core/gprpp/dual_ref_counted_test.cc:30:20: note: in instantiation of template class 'grpc_core::DualRefCounted<grpc_core::testing::(anonymous namespace)::Foo>' requested here
class Foo : public DualRefCounted<Foo> {
                   ^
test/core/gprpp/dual_ref_counted_test.cc:30:7: note: definition of 'grpc_core::testing::(anonymous namespace)::Foo' is not complete until the closing '}'
class Foo : public DualRefCounted<Foo> {
      ^
```

So I don't think this is going to work anyway.

Reply:
I was hoping to improve the error message quality if someone tried to use this without CRTP, but it looks like it's not going to work. Ok.